### ISSUE TITLE
3 Events

### ISSUE THREAD

**Issue by splitice** [tunnelvisionlabs/NOpenCL#7]
OS: Linux
Platform: Debian w/ Mono
Architecture: armv71

When I pass in 3 Events to wait on I get a crash. When investigating with Valgrind on Linux I found what looks to be signs of an error in the Safe Handle Array code.

3 events (no crash but same error):

```
==28599== Invalid write of size 4
==28599==    at 0x6ABD838: System_Runtime_InteropServices_Marshal_WriteInt32_intptr_int_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28599==    by 0x6ABD9B3: System_Runtime_InteropServices_Marshal_WriteIntPtr_intptr_int_intptr (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28599==    by 0x169146CF: ???
==28599==  Address 0xd481b7c is 0 bytes after a block of size 4 alloc'd
==28599==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28599==
==28599== Invalid read of size 4
==28599==    at 0xF73B986: mcl_entrypoints_valid_event_list (in /usr/lib/libOpenCL.so)
==28599==  Address 0xd481b7c is 0 bytes after a block of size 4 alloc'd
==28599==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28599==
==28599== Invalid read of size 4
==28599==    at 0xF73BFBC: mcl_wait_for_events (in /usr/lib/libOpenCL.so)
==28599==  Address 0xd481b7c is 0 bytes after a block of size 4 alloc'd
==28599==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28599==
==28599== Invalid read of size 4
==28599==    at 0xF73BFD8: mcl_wait_for_events (in /usr/lib/libOpenCL.so)
==28599==  Address 0xd481b7c is 0 bytes after a block of size 4 alloc'd
==28599==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28599==
==28599== Invalid read of size 4
==28599==    at 0x6ABC8E4: System_Runtime_InteropServices_Marshal_ReadInt32_intptr_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28599==    by 0x6ABCA53: System_Runtime_InteropServices_Marshal_ReadIntPtr_intptr_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28599==    by 0x16914943: ???
==28599==  Address 0xd481b7c is 0 bytes after a block of size 4 alloc'd
==28599==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
```

2 events (no crash but same error):
```
==28386== Invalid write of size 4
==28386==    at 0x6ABD838: System_Runtime_InteropServices_Marshal_WriteInt32_intptr_int_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28386==    by 0x6ABD9B3: System_Runtime_InteropServices_Marshal_WriteIntPtr_intptr_int_intptr (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28386==    by 0x16A1478F: ???
==28386==  Address 0xdd184e4 is 0 bytes after a block of size 4 alloc'd
==28386==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28386==
==28386== Invalid read of size 4
==28386==    at 0xF83B986: mcl_entrypoints_valid_event_list (in /usr/lib/libOpenCL.so)
==28386==  Address 0xdd184e4 is 0 bytes after a block of size 4 alloc'd
==28386==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28386==
==28386== Invalid read of size 4
==28386==    at 0xF83BFBC: mcl_wait_for_events (in /usr/lib/libOpenCL.so)
==28386==  Address 0xdd184e4 is 0 bytes after a block of size 4 alloc'd
==28386==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28386==
==28386== Invalid read of size 4
==28386==    at 0xF83BFD8: mcl_wait_for_events (in /usr/lib/libOpenCL.so)
==28386==  Address 0xdd184e4 is 0 bytes after a block of size 4 alloc'd
==28386==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
==28386==
==28386== Invalid read of size 4
==28386==    at 0x6ABC8E4: System_Runtime_InteropServices_Marshal_ReadInt32_intptr_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28386==    by 0x6ABCA53: System_Runtime_InteropServices_Marshal_ReadIntPtr_intptr_int (in /usr/lib/mono/aot-cache/arm/mscorlib.dll.so)
==28386==    by 0x16A14A03: ???
==28386==  Address 0xdd184e4 is 0 bytes after a block of size 4 alloc'd
==28386==    at 0x483E394: malloc (in /usr/lib/valgrind/vgpreload_memcheck-arm-linux.so)
```

**Comment by splitice** [comment]
Perhaps this: https://github.com/splitice/NOpenCL/commit/e4767b753c20196ba046ac62734d0787d1d51172
?

**Comment by sharwell** [comment]
@splitice Nice catch

**Comment by sharwell** [comment]
@splitice Want to send a pull request with that change? I don't mind doing it myself but since you found it I want to give you more opportunity to take the credit for it. 😉

**Comment by splitice** [comment]
I'd be happy to.

Out of curiosity, any thoughts on my other two commits that add a KernelWrapper & MSBuild BuildTask similar to OpenCL.Net, Not 100% complete yet (works, just need to commit the target & make it handle OpenCL build errors). Is that something you would be interested in accepting upstream?

**Comment by sharwell** [comment]
@splitice I recommend filing new issues for the other questions. The biggest win that I'd love to see fixed is #9. Ever since I moved my development machine to Hyper-V I haven't been able to do much with this project. 😦
